gem/views.py

```diff
@@ -55,14 +64,172 @@ def form_valid(self, form):
+
+        if self.request.session['forgot_password_attempts'] >= 5:
+            # GEM-195 implemented a 10 min session timeout, so effectively
+            # the user can only try again once their anonymous session expires.
+            # If they make another request within the 10 min time window the
+            # expiration will be reset to 10 mins in the future.
+            # This is obviously not bulletproof as an attacker could simply
+            # not send the session cookie to circumvent this.
+            form.add_error(None, 'Too many attempts. Please try again later.')
+            return self.render_to_response({'form': form})
+
+        username = form.cleaned_data['username']
+        random_security_question_idx = self.request.session[
+            'random_security_question_idx'
+        ]
+        random_security_question_answer = form.cleaned_data[
+            'random_security_question_answer'
+        ]
+
+        # TODO: consider moving these checks to GemForgotPasswordForm.clean()
+        # see django.contrib.auth.forms.AuthenticationForm for reference
+        try:
+            user = User.objects.get_by_natural_key(username)
+        except User.DoesNotExist:
+            self.request.session['forgot_password_attempts'] += 1
+            form.add_error(
+                'username', 'The username that you entered appears to be '
+                            'invalid. Please try again.')
+            return self.render_to_response({'form': form})
+
+        if not user.is_active:
+            form.add_error('username', 'This account is inactive.')
+            return self.render_to_response({'form': form})
+
+        is_answer_correct = False
+        if random_security_question_idx == 0:
+            is_answer_correct = \
+                user.gem_profile.check_security_question_1_answer(
+                    random_security_question_answer
+                )
+        elif random_security_question_idx == 1:
+            is_answer_correct = \
+                user.gem_profile.check_security_question_2_answer(
+                    random_security_question_answer
+                )
+        else:
+            logging.warn('Unhandled security question index')
+
+        if not is_answer_correct:
+            self.request.session['forgot_password_attempts'] += 1
+            form.add_error('random_security_question_answer',
+                           'Your answer to the security question was invalid. '
+                           'Please try again.')
+            return self.render_to_response({'form': form})
+
+        # NB: NOT safe if cookie-based sessions are used (with cookie-based
+        # sessions the session data is stored in the cookie itself and is
+        # NOT encrypted!)
+        # See https://docs.djangoproject.com/en/1.9/topics/http/sessions/
+        # #configuring-the-session-engine
+        # TODO: consider generating a reset token for the user and redirecting
```

Comment:
Agreed. Please use token and redirect instead. See this example from [django email reset](https://github.com/django/django/blob/master/django/contrib/auth/forms.py#L274)